theano/gpuarray/GpuMaxAndArgmax.py

```diff
@@ -0,0 +1,209 @@
+             * in the input, so that the 2nd dimenson of the matrix will be reduced. */
+            size_t total_size = 1;
+            for(size_t i = 0; i < %(name)s_input_ndim; ++i) {
+                total_size *= PyGpuArray_DIM(%(X)s, i);
+            }
+            size_t newdims[2] = {1, total_size};
+            %(name)s_input = new GpuArray;
+            if(GA_NO_ERROR !=
+                GpuArray_reshape(%(name)s_input, &%(X)s->ga, 2, newdims, GA_ANY_ORDER, 0)
+            ) {
+                %(fail)s
+            }
+            %(name)s_redux_len = 1;
+            %(name)s_axes_to_reduce[0] = 1;
+        } else {
+            %(name)s_output_dims = new size_t[%(name)s_output_ndim];
+            if(%(name)s_redux_len == 1) {
+                for(unsigned i = 0; i < %(name)s_axes_to_reduce[0]; ++i) {
+                    %(name)s_output_dims[i] = PyGpuArray_DIM(%(X)s, i);
+                }
+                for(unsigned i = %(name)s_axes_to_reduce[0] + 1; i < %(name)s_input_ndim; ++i) {
+                    %(name)s_output_dims[i-1] = PyGpuArray_DIM(%(X)s, i);
+                }
+            } else {
+                int64_t current_input_pos = -1;
+                int64_t current_output_pos = -1;
+                for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
+                    for(++current_input_pos; current_input_pos < %(name)s_axes_to_reduce[i]; ++current_input_pos) {
+                        %(name)s_output_dims[++current_output_pos] = PyGpuArray_DIM(%(X)s, current_input_pos);
+                    }
+                }
+                for(++current_input_pos; current_input_pos < %(name)s_input_ndim; ++current_input_pos) {
+                    %(name)s_output_dims[++current_output_pos] = PyGpuArray_DIM(%(X)s, current_input_pos);
+                }
+            }
+        }
+        if(theano_prep_output(&%(max)s, %(name)s_output_ndim, %(name)s_output_dims, %(max_typecode)s, GA_C_ORDER, %(X)s->context)) {
+            %(fail)s
+        }
+        if(theano_prep_output(&%(argmax)s, %(name)s_output_ndim, %(name)s_output_dims, %(argmax_typecode)s, GA_C_ORDER, %(X)s->context)) {
+            %(fail)s
+        }
+        if(GA_NO_ERROR !=
+            GpuArray_maxandargmax(&%(max)s->ga, &%(argmax)s->ga, %(name)s_input, %(name)s_redux_len, %(name)s_axes_to_reduce)
+        ) {
+            %(fail)s
+        }
+        """
+        if theano.config.gpuarray.sync:
+            ret += """
+            GpuArray_sync(&%(max)s->ga);
+            GpuArray_sync(&%(argmax)s->ga);
+            """
+        return ret % {'X': input_names[0], 'axes': input_names[1], 'max': output_names[0], 'argmax': output_names[1],
+                      'axes_ctype': axes_ctype, 'max_typecode': max_typecode, 'argmax_typecode': argmax_typecode,
+                      'name': name, 'fail': sub['fail']}
+
+    def c_code_cleanup(self, node, name, inputs, outputs, sub):
+        return """
+        delete[] %(name)s_output_dims;
```

Comment:
In C++, deleting a NULL pointer does normally nothing ! http://en.cppreference.com/w/cpp/language/delete

But as you required, I will remove any C++ syntax from the code !